lib/yggdrasil/player.ex

```diff
@@ -12,9 +13,16 @@ defmodule Yggdrasil.Player do
     Supervisor.add_player user_id, [owner_pid, push_msg]
   end
 
-  def run_cmd(user_id, cmd = %Message{type: :command}) do
-    player_pid = Registry.get_player user_id
-    GenServer.cast(player_pid, {:run_cmd, cmd})
+  def run_cmd(user_id, cmd) do
+    # horrible terrible temporary implementation
+    # room would eventually do all of this
+    ctxt = %Yggdrasil.Room.Context{ player: user_id }
+    ctxt = Command.execute cmd, ctxt
+
+    Enum.each ctxt.actions,
+      fn {:notify, player, message} ->
+        notify(player, message)
+      end
```

Comment:
Ah ok, yeah so I don't see why a context could not be validated first right? In the example I provided the intent of a command gets collapsed down to a list of actions which the room process receives.

So then the room could do this

```elixir
# room module
def run(ctx) do
  # I would map over the ctc.actions, each one either giving {:ok, action}, or {:error, msg}
  # then check to see if there are any errors before performing the actions.
  ctx = validate(ctx)
end
```
However I now understand what you are doing, and good with that as well. The command's context would get the room struct then the room calls execute and each action in the command's pipeline would probably needs to have some way of effectively not produce anything if there are errors. Ectos changesets do that for example with the valid? key. If something is invalid it sets this and the next function that touches the changeset will effectively be bypassed.

So yeah I'm good with that as well. I think my thoughts were more like it's ok if the context is not valid, it's describing the intent and just being built then it gets validated then applied. This way the commands don't have to know anything about the room just how to build the right forms of the actions to apply.

Anyway sorry, I'm good with what you are doing what you just said made perfect sense to me and see where you are headed with it.